### ISSUE TITLE
Smallcaps Cheatsheet

### ISSUE THREAD

**Issue by erights** [ocapn/ocapn#91]
As promised at the ocapn meeting, here is the small summary of the Smallcaps encoding:

https://github.com/endojs/endo/blob/master/packages/marshal/docs/smallcaps-cheatsheet.md

What should we do at this ocapn repository to reflect this?

**Comment by jar398** [comment]
I don't understand the column headings 

"Passable value | OCapn name"

The first column does not contain values, it looks like it contains type names?  But in what language - intended to be human readable and smallcaps/syrup-neutral maybe?

The second column is confusing to me because I don't know what "OCapn" means.

**Comment by jar398** [comment]
What is a "special string"?  (This must be described somewhere, but it seems a short description would do well in this document.)

**Comment by erights** [comment]
> Passable value

Since the cheatsheet is in the endo repo and is intended to help understand the endo implementation of OCapn, the descriptions in this column have a bias towards JS and our mapping of OCapn values to JS values. For OCapn purposes aside from help understanding the code of the existing JS implementation, you can start at the second column...

> OCapn name

> The second column is confusing to me because I don't know what "OCapn" means.

I take the thread starting at @zenhack 's May 15 comment at
https://github.com/ocapn/ocapn/issues/5#issuecomment-1549012122

to represent the state of agreement OCapn has already achieved on abstract syntax / data model. I hope we can continue to make progress from this as an agreed baseline.

> What is a "special string"? (This must be described somewhere, but a short description would do well in this document.)

Already defined at the fourth bullet on that page

> * Special strings begin with any of the `!"#$%&'()*+,-` characters.

**Comment by erights** [comment]
>  for an appropriate definition of 'string')

At https://github.com/endojs/endo/issues/1587 I list 6 changes we intend to make to bring the Agoric implementations of OCapn into alignment with the agreed OCapn. In our phone call I only remembered 3 of these. Strings are number 5 on that list, also discussed at https://github.com/ocapn/ocapn/issues/47 .

**Comment by cwebber** [comment]
First of all, this cheat sheet is great!  Thank you @erights!

Regarding putting it in OCapN's repositories, I suggest we make it symmetric to what we have done with Syrup: instead of putting it in `ocapn/ocapn`, we start a `ocapn/smallcaps` repository (same as we have with `ocapn/syrup`) and put the cheat sheet there as a starting point and link to implementations.   Speaking of, @tsyesika has something to say about that...

**Comment by tsyesika** [comment]
I have implemented a smallcaps encoder in guile for us to use in goblins. The implementation is here:

```scheme
(define (smallcaps-encode value)
  (define (encode-value value)
    (match value
      [(? string?)
       (cond [(string-null? value) value]
             [(restricted-char? (string-ref value 0)) (string-append "!" value)]
             [else value])]
      [(? symbol?)
       (string-append "%" (symbol->string value))]
      [(? boolean?) value]
      [(? desc:import-object?)
       (string-append "$" (number->string (desc:import-object-pos value)))]
      [(? desc:import-promise?)
       (string-append "&" (number->string (desc:import-promise-pos value)))]
      ;; Two bottom types
      [(? unspecified?) "#undefined"]
      [(? zilch?) 'null]
      ;; Numeric types
      [(and (? number?) (? inf?))
       (if (negative? value) "#-Infinity" "#Infinity")]
      [(and (? number?) (? nan?)) "#NaN"]
      [(? integer?)
       (string-append (if (positive? value) "+" "-")
                      (number->string value))]
      [(and (? number?) (? inexact?)) value]
      [($ <tagged> label value)
       `(("#tag" . ,(encode-value label))
         (payload . ,(encode-value value)))]
      [(? ghash?)
       ;; Conver the ghash to an alist while checking the keys are all strings.
       (ghash-fold
        (lambda (key value alist)
          (if (string? key)
              (cons (cons key value) alist)
              (error "Smallcaps can only encode keys of type string:" key)))
        '()
        value)]
      ;; No support for binary data
      ;;[(? bytevector?) (error "No support for binary data type in smallcaps (yet)")]
      [(? bytevector? bv)
       (encode-value
        (make-tagged 'binrary-data (base64-encode bv)))]
      [(? list?) (list->vector (map encode-value value))]))
  (scm->json-string (encode-value value)))
```

This implementation does actually include binary data serialization as a "tagged" value with a base64 encoded payload. This is not something I'm suggesting smallcaps actually does, it's just a stop gap measure until a binary data type is added.

**Comment by jar398** [comment]
This is helpful. To be complete (e.g. for round trip testing) we'd need a decoder too, yes?

**Comment by erights** [comment]
Yes, this is great! I think I'll try to implement the same stopgap for binary data for now, so it is not an impediment to our interoperability experiments.

I want to acknowledge that the issue of canonical decimal expansion of float64s is something I did not think of in our earlier morning meeting when I thought canonical smallcaps would be easy to spec and implement. (attn @gibson042 ) What does your code currently do with rendering float64s into JSON decimal floating point numbers?

misspelling: "binrary" vs "binary"?

The smallcaps cheatsheet does specify that the value of a tag be recursively encoded/decoded, which indeed the Agoric implementation does. The OCapn agreement starting with @zenhack's May 15 message agrees that the value of a tag be a string. IIUC, the tag encoding I see above would decode to a symbol, which the Agoric implementation, I believe, would correctly reject as non-conformant.

I couldn't tell what your code above does with -0

**Comment by erights** [comment]
IIUC, you're using the slot index as the c-list index for that context's c-list, rather than a per-encoding slots array. Is that right?
